app/src/main/kotlin/de/rochefort/childmonitor/MonitorService.kt

```diff
@@ -0,0 +1,270 @@
+        // Display a notification about us starting.  We put an icon in the status bar.
+        createNotificationChannel()
+        val n = buildNotification()
+        val foregroundServiceType = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) ServiceInfo.FOREGROUND_SERVICE_TYPE_MICROPHONE else 0 // Keep the linter happy
+        ServiceCompat.startForeground(this, ID, n, foregroundServiceType)
+        ensureMonitorThread()
+        return START_REDELIVER_INTENT
+    }
+
+    private fun ensureMonitorThread() {
+        var mt = this.monitorThread
+        if (mt != null && mt.isAlive) {
+            return
+        }
+        val currentToken = Any()
+        this.connectionToken = currentToken
+        mt = Thread {
+            while (this.connectionToken == currentToken) {
+                try {
+                    ServerSocket(this.currentPort).use { serverSocket ->
+                        this.currentSocket = serverSocket
+                        // Store the chosen port.
+                        val localPort = serverSocket.localPort
+
+                        // Register the service so that parent devices can
+                        // locate the child device
+                        registerService(localPort)
+                        serverSocket.accept().use { socket ->
+                            Log.i(TAG, "Connection from parent device received")
+
+                            // We now have a client connection.
+                            // Unregister so no other clients will
+                            // attempt to connect
+                            unregisterService()
+                            serviceConnection(socket)
+                        }
+                    }
+                } catch (e: Exception) {
+                    if (this.connectionToken == currentToken) {
+                        // Just in case
+                        this.currentPort++
+                        Log.e(TAG, "Failed to open server socket. Port increased to $currentPort", e)
+                    }
+                }
+            }
+        }
+        this.monitorThread = mt
+        mt.start()
+    }
+
+    private fun registerService(port: Int) {
+        val serviceInfo = NsdServiceInfo()
+        serviceInfo.serviceName = "ChildMonitor on " + Build.MODEL
+        serviceInfo.serviceType = "_childmonitor._tcp."
+        serviceInfo.port = port
+        this.registrationListener = object : RegistrationListener {
+            override fun onServiceRegistered(nsdServiceInfo: NsdServiceInfo) {
+                // Save the service name.  Android may have changed it in order to
+                // resolve a conflict, so update the name you initially requested
+                // with the name Android actually used.
+                val serviceName = nsdServiceInfo.serviceName
+                Log.i(TAG, "Service name: $serviceName")
+                val ma = monitorActivity
```

Comment:
We can get rid of this temp variable with another let block.

Reply:
Done